File: app/src/main/java/com/absinthe/kage/device/cmd/DeviceRotationCommand.java
package com.absinthe.kage.device.cmd;

import android.content.pm.ActivityInfo;
import android.content.res.Configuration;

import com.absinthe.kage.BaseActivity;
import com.absinthe.kage.connect.protocol.IpMessageConst;
import com.absinthe.kage.connect.protocol.IpMessageProtocol;
import com.absinthe.kage.device.Command;
import com.absinthe.kage.device.CommandBuilder;
import com.absinthe.kage.device.client.Client;
import com.absinthe.kage.manager.ActivityStackManager;

public class DeviceRotationCommand extends Command {

    public static final int TYPE_LAND = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE;
    public static final int TYPE_PORT = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT;

    public static final int LENGTH = 2;
    public int flag = TYPE_LAND;

    public DeviceRotationCommand() {
        cmd = IpMessageConst.DEVICE_ROTATION;
    }

    @Override
    public String pack() {
        return new CommandBuilder()
                .with(this)
                .append(String.valueOf(flag))
                .build();
    }

    @Override
    public void doWork(Client client, String received) {
        if (parseReceived(received)) {
            BaseActivity topActivity = ActivityStackManager.INSTANCE.getTopActivity();
            if (topActivity != null) {
                topActivity.setRequestedOrientation(flag);
            }
        }
    }

    @Override
    public boolean parseReceived(String received) {
        String[] splits = received.split(IpMessageProtocol.DELIMITER);
        if (splits.length == LENGTH) {
            try {
                flag = Integer.parseInt(splits[1]);
                if (flag == Configuration.ORIENTATION_LANDSCAPE) {
                    flag = TYPE_LAND;
                } else {
                    flag = TYPE_PORT;
                }
            } catch (NumberFormatException e) {
                e.printStackTrace();
                return false;
            }
            return true;
        } else {
            return false;
        }
    }
}

File: app/src/main/kotlin/com/absinthe/kage/device/cmd/DeviceRotationCommand.kt
package com.absinthe.kage.device.cmd

import android.content.pm.ActivityInfo
import android.content.res.Configuration
import com.absinthe.kage.connect.protocol.IpMessageConst
import com.absinthe.kage.connect.protocol.IpMessageProtocol
import com.absinthe.kage.device.Command
import com.absinthe.kage.device.CommandBuilder
import com.absinthe.kage.device.client.Client
import com.absinthe.kage.manager.ActivityStackManager.topActivity

class DeviceRotationCommand : Command() {

    var flag = TYPE_LAND

    override fun pack(): String {
        return CommandBuilder()
                .with(this)
                .append(flag.toString())
                .build()
    }

    override fun doWork(client: Client, received: String) {
        if (parseReceived(received)) {
            val topActivity = topActivity

            if (topActivity != null) {
                topActivity.requestedOrientation = flag
            }
        }
    }

    override fun parseReceived(received: String): Boolean {
        val splits = received.split(IpMessageProtocol.DELIMITER).toTypedArray()

        return if (splits.size == LENGTH) {
            try {
                flag = splits[1].toInt()
                flag = if (flag == Configuration.ORIENTATION_LANDSCAPE) {
                    TYPE_LAND
                } else {
                    TYPE_PORT
                }
            } catch (e: NumberFormatException) {
                e.printStackTrace()
                return false
            }
            true
        } else {
            false
        }
    }

    companion object {
        const val TYPE_LAND = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
        const val TYPE_PORT = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        const val LENGTH = 2
    }

    init {
        cmd = IpMessageConst.DEVICE_ROTATION
    }
}